[Woo POS][Barcodes] Implement barcode scan heuristics

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds scan heuristics to ensure we're handling input from a scanner, not an external keyboard.

I've matched the implementation in https://github.com/woocommerce/woocommerce-android/pull/14162

### Heuristics
- No more than 100ms between keystrokes
- No more than 1.5s to complete the scan – at the end of the time, we consider it a scan in case a terminator isn't sent
- Barcode must have at least 4 characters

Beyond the Android implementation, I added special handling for iOS sending us control codes.

Without this, if the scanner's configured with `CRLF` as the terminator, we get `UIKeyInputDownArrow` as a barcode input after a successful scan – obviously this fails.

We have a 1.5s timeout, at which point we try to process input. I've opened a discussion about this because I don't think it's necessarily the best thing for the product, but for the now we match Android. We may decide to ignore scans if they hit the timeout, instead.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Enable the `pointOfSaleBarcodeScanningi1` feature flag

Open Point of Sale and test barcodes using a hardware scanner. They should work as expected. 

Try with the terminator turned off – scanning should still work, it's just slower because of the timeout.

Very long barcodes may not complete in time – these will cause errors, possibly multiple, rather than being ignored.

Try with different combinations of `\r \n` as terminators – they should work.

Try typing on an external keyboard – it should be difficult to get that to register as a scan.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

I've tested with an iPad Air on iOS 17.7.2, a Star BSH-20B, and an external keyboard.

Note that the barcode scan simulator doesn't test this – it bypasses the `HIDBarcodeParser`, and it'd be a bit of a challenge to make it work through that now, as we need to get the keypresses, not just the characters to do it properly.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/34901256-2978-4007-8dbb-982022305697

There's a lot of me rambling at the end about keyboard layouts. I tested it, and the best case for QR codes is to set the scanner up in US English, and the Hardware Keyboard layout in iOS Settings in US English, and then kill and restart the app for good measure. If you do that, it starts giving the expected barcodes!

Fortunately this doesn't make much difference for normal barcodes, the numbers are pretty universal... though we'll have to check it in Arabic when we expand POS internationally.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
Only handle barcodes which may be from a scanner

joshheald committed:
Add tests for scanning heuristics

joshheald committed:
Fix test for processing scan after max scan time

joshheald committed:
Ignore non-character keys in scans

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Ensure special characters in barcodes are parsed

Previous use of `charactersIgnoringModifiers` was convenient, but meant that anything which should be considered with a modifier would just use the raw keystroke, e.g. `2` instead of `@`, which could break some edge case workflows.

joshheald commented on WooCommerce/Classes/POS/Presentation/Barcode Scanning/HIDBarcodeParser.swift:
```diff
@@ -9,27 +22,153 @@ final class HIDBarcodeParser {
+        }
+        lastKeyPressTime = currentTime
+
+        let character = key.characters
```
Previous use of `charactersIgnoringModifiers` as passed in from the container was convenient, but meant that anything which should be considered with a modifier would just use the raw keystroke, e.g. `2` instead of `@`, which could break some edge case workflows.

joshheald commented on WooCommerce/Classes/POS/Presentation/Barcode Scanning/HIDBarcodeParser.swift:
```diff
@@ -9,27 +22,153 @@ final class HIDBarcodeParser {
+        scanTimer = nil
+    }
+
+    private func startScanTimer() {
+        scanTimer?.invalidate()
+        scanTimer = timerFactory.createTimer(interval: configuration.maximumScanTime, repeats: false) { [weak self] in
+            guard let self = self else { return }
+            self.processScan()
+        }
+    }
```
Timers are unpredictable from tests, this lets us inject them

joshheald commented on WooCommerce/Classes/POS/Presentation/Barcode Scanning/BarcodeScannerContainer.swift:
```diff
@@ -103,7 +104,7 @@ class BarcodeScannerHostingController: UIHostingController<EmptyView> {
         /// It's better practice for barcode scanning to only consider the presses when they end.
         for press in presses {
-            guard let key = press.key?.charactersIgnoringModifiers else { continue }
+            guard let key = press.key else { continue }
```
We pass the raw key now, to allow us to test the modifier/special key handling.

Note that the previous use of `charactersIgnoringModifiers` meant we used the raw keystroke, e.g. `2` instead of `@`, which could break some edge case workflows with 2D codes.

joshheald committed:
Remove unnecessary import

staskus commented on WooCommerce/Classes/POS/Presentation/Barcode Scanning/HIDBarcodeParser.swift:
```diff
@@ -9,27 +22,153 @@ final class HIDBarcodeParser {
         }
     }
 
+    private let excludedKeys: [UIKeyboardHIDUsage] = [
```
Is there a practical value in excluding keys? As far as I understand, if they arrived from the barcode scanner, something would be misaligned with the keyboard & barcode scanner configuration?

joshheald replied:
This is what prompted me to do it – a single scan can do this without the exclusions:

https://github.com/user-attachments/assets/44f0a115-33f8-4e6d-a641-13f10271ca6d

Barcode used: https://barcode-maker.com/Qrcode?data=https%3A%2F%2Fwoocommerce.com12345!%40%23%24%25%5E

The extra `UIKeyInputDownArrow` happens when you configure the Star barcode scanner to send CRLF as the terminating character, instead of just CR. Those are the two options presented in the quick start guide.

Perhaps this list goes too far... it's the key positions (on a US English keyboard) that we're excluding, not the characters themselves. But it seems that iOS provides the characters for these keys as a name of the key position, at least in some cases.

I've tried without the guard for this, and the shift key (for example) doesn't get sent by our scanner... but it could be, in theory, and we wouldn't want to separately spell it out in a barcode read if it was.

staskus replied:
Got it. If we do support URLs, we need to support `keypadEqualSign`, so yes, maybe the exclusion list could be shortened. However, if we only support `GTIN`, then it's less of a concern.

joshheald replied:
Good spot! I'll remove that one from the list.

staskus commented on WooCommerce/WooCommerceTests/POS/Presentation/Barcode Scanning/HIDBarcodeParserTests.swift:
```diff
@@ -13,114 +13,320 @@ struct HIDBarcodeParserTests {
+        parser.processKeyPress(MockUIKey(character: "1"))
+        parser.processKeyPress(MockUIKey(character: "2"))
+        parser.processKeyPress(MockUIKey(character: "3"))
+        Thread.sleep(forTimeInterval: 0.06) // Just over maximumInterCharacterTime
```
I hope this doesn't introduce any flakiness into the tests. 

If we want to make `HIDBarcodeParser` testable without artificial timeouts, we could make not only the timer injectable (`DefaultTimerFactory`) but also the current time injectable (e.g `DefaultTimeProvider()`). In tests, we could manually move the current time past the `maximumScanTime` or `maximumInterCharacterTime`. 

Example:

```swift
protocol TimeProvider {
    func now() -> Date
}

struct DefaultTimeProvider: TimeProvider {
    func now() -> Date {
        Date()
    }
}


private class MockTimeProvider: TimeProvider {
    private var currentTime: Date
    
    init(startTime: Date = Date()) {
        self.currentTime = startTime
    }
    
    func now() -> Date {
        currentTime
    }
    
    func advance(by interval: TimeInterval) {
        currentTime = currentTime.addingTimeInterval(interval)
    }
}
```

```swift
    func testSlowTyping() {
        var scannedCodes: [String] = []
        let timeProvider = MockTimeProvider()
        let configuration = HIDBarcodeParserConfiguration(
            terminatingStrings: ["\r"],
            minimumBarcodeLength: 4,
            maximumScanTime: 0.3,
            maximumInterCharacterTime: 0.05
        )
        let parser = HIDBarcodeParser(
            configuration: configuration,
            onScan: { code in
                scannedCodes.append(code)
            },
            timeProvider: timeProvider
        )

        // Simulate slow typing
        parser.processKeyPress(MockUIKey(character: "1"))
        parser.processKeyPress(MockUIKey(character: "2"))
        parser.processKeyPress(MockUIKey(character: "3"))
        timeProvider.advance(by: 0.06) // Just over maximumInterCharacterTime
        parser.processKeyPress(MockUIKey(character: "4"))
        parser.processKeyPress(MockUIKey(character: "\r"))

        #expect(scannedCodes.isEmpty)
    }
 ````

joshheald replied:
Good point. I'll have a poke at this before I merge

joshheald replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/15768/commits/df489d56450425d26cf2288707263ddcbf25ee38

staskus reviewed (approved):
I tested as best as I could've with the external keyboard. It's much harder to get it to work, I need to essentially push the keys at once.

I have a suggestion regarding tests, and uncoupling them from waiting times.

I'm also not sure of the value of this long list of excluded keys.

joshheald committed:
Inject TimeProvider to avoid flaky tests.

joshheald committed:
Fix broken test

joshheald committed:
Set maximum time between keys to 200ms

This should accommodate slower scanners

joshheald committed:
Remove maximum barcode scan time

We have not found examples of scanners which don’t send `\r\n` style terminators at the end of a scan. Processing a scan after a delay risks being more confusing and makes the code more complex, so we’ve decided not to do it any more.

This commit removes all the code related to the 1.5s timeout for scans

joshheald committed:
Remove keypad equal sign from exclusions
